File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        HashSet<Character> seen = new HashSet<Character>();
        int windowstart = 0;
        int max = Integer.MIN_VALUE;
        
        for(int windowend = 0; windowend < s.length(); windowend++){
               if(seen.contains(s.charAt(windowend)) == false ){
                   seen.add(s.charAt(windowend)); 
                 }
               else{
                    while(seen.contains(s.charAt(windowend)) && windowend - windowstart >= 1){
                         seen.remove(s.charAt(windowstart));  
                         windowstart++;  
                     }
                   seen.add(s.charAt(windowend));
               }
               max = Math.max(windowend - windowstart + 1, max);             
        }
        if (max == Integer.MIN_VALUE){return 0;}
        return max;
    }      
}
       

        

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        dupes = set()
        start = 0
        end = 0 
        max_size = 0
        while end < len(s):

            while len(dupes)!=0 and s[end] in dupes:
                dupes.remove(s[start])
                start+=1

            max_size =  max(max_size, end-start+1)
            dupes.add(s[end])  
            end+=1  
        return max_size    